diff --git a/config/Makefile.linux b/config/Makefile.linux
index 8d154cd..81940ca 100644
--- a/config/Makefile.linux
+++ b/config/Makefile.linux
@@ -1,25 +1,7 @@
 # Linux Epics IOC system build rules
 
-
-###
-#  if EPICS environment variables defined, use those
-#  - otherwise set to old defaults for now
-ifeq (${EPICS_BASE},)
-EPICS_BASE = /opt/rtapps/epics/base
-EPICS_SEQ = /opt/rtapps/epics/modules/sncseq
-EPICS_EXTENSIONS = /opt/rtapps/epics/extensions
-endif
-ifeq (${EPICS_HOST_ARCH},)
-EPICS_HOST_ARCH = linux-x86_64
-endif
-
-EPICS_BASE_PATH := $(EPICS_BASE)
-EPICS_SEQ_PATH := $(EPICS_SEQ)
-EPICS_EXT_PATH := $(EPICS_EXTENSIONS)
-EPICS_ARCH := $(EPICS_HOST_ARCH)
-
-EPICS_BIN_PATH := $(EPICS_BASE)/bin/$(EPICS_ARCH)
-EPICS_LIB_PATH := $(EPICS_BASE)/lib/$(EPICS_ARCH)
+EPICS_VERSION := $(shell pkg-config --modversion epics-base)
+EPICS_BIN_PATH := $(shell pkg-config --variable=bindir epics-base)
 
 # Sequencer synchronization; to fix broken seq package
 SRC += $(RCG_SRC_DIR)/src/drv/sync.c
@@ -32,9 +14,7 @@ PREPROCESS = /usr/bin/gcc -x c -E -P
 
 # Standard C and C++ compiler and C preprocessor flags
 CFLAGS += $(EXTRA_CFLAGS)
-CFLAGS += -I$(EPICS_BASE_PATH)/include/os/Linux -I$(EPICS_BASE_PATH)/include/compiler/gcc -I$(EPICS_BASE_PATH)/include 
-CFLAGS += -I$(EPICS_SEQ_PATH)/include
-CFLAGS += -I$(EPICS_EXT_PATH)/include
+CFLAGS += $(shell pkg-config --cflags epics-base)
 CFLAGS += -I/usr/lib/rmnet -I/usr/lib/rfm2g
 CFLAGS += -Isrc/include
 CFLAGS += -I$(RCG_SRC_DIR)/src/include -I$(RCG_SRC_DIR)/src/include/drv
@@ -58,7 +38,7 @@ SNCFLAGS += +a -c -w
 #SNCFLAGS += -e
 
 # Sequencer file (.st) compiler
-SNC = $(EPICS_SEQ_PATH)/bin/$(EPICS_ARCH)/snc $(SNCFLAGS)
+SNC = $(EPICS_BIN_PATH)/snc $(SNCFLAGS)
 
 # registration code generator
 REGGEN = $(EPICS_BIN_PATH)/registerRecordDeviceDriver.pl
@@ -68,12 +48,10 @@ FMSEQFILES = $(RCG_SRC_DIR)/src/epics/util/fmseq.pl $(RCG_SRC_DIR)/src/epics/uti
 
 # Standard Epics IOC server library flags
 LIBFLAGS += -g
-LIBFLAGS += -L$(EPICS_LIB_PATH)
-LIBFLAGS += -L$(EPICS_SEQ_PATH)/lib/$(EPICS_ARCH)
-LIBFLAGS += -L$(EPICS_EXT_PATH)/lib/$(EPICS_ARCH)
+LIBFLAGS += $(shell pkg-config --libs epics-base)
 
 # test for 3.15 libraries, otherwise use 3.14
-ifneq ("$(wildcard $(EPICS_LIB_PATH)/libdbCore.a)","")
+ifneq ($(wordlist 1,2,$(subst '.',' ',$(EPICS_VERSION))), (3 15))
 LIBFLAGS += -lseq -lpv -ldbRecStd -ldbCore
 LIBFLAGS += -lca -lCom
 else
diff --git a/src/epics/util/feCodeGen.pl b/src/epics/util/feCodeGen.pl
index 27265c6..830034a 100755
--- a/src/epics/util/feCodeGen.pl
+++ b/src/epics/util/feCodeGen.pl
@@ -2386,7 +2386,6 @@ sub createEpicsMakefile {
 	  print OUTME "EXTRA_CFLAGS += -DUSE_SYSTEM_TIME=1\n";
 	}
 	print OUTME "\n";
-	print OUTME "LIBFLAGS += -lezca\n";
 	if($systemName eq "sei" || $useFIRs)
 	{
 	print OUTME "EXTRA_CFLAGS += -DFIR_FILTERS\n";
diff --git a/src/epics/util/skeleton.st b/src/epics/util/skeleton.st
index 5e4fe7b..c9dba74 100644
--- a/src/epics/util/skeleton.st
+++ b/src/epics/util/skeleton.st
@@ -14,7 +14,6 @@ program %SEQUENCER_NAME%
 %% #include <errno.h>
 %% #include <epicsTime.h>
 %% #include <cadef.h>
-%% #include <ezca.h>
 %% #include <alarm.h>                                                    /* MA */
 %% #include <sys/types.h>                                                /* MA */
 %% #include <sys/stat.h>                                                 /* MA */
